### PR TITLE
Add documentation endpoint to lang-server

### PR DESCRIPTION
**Pull request by sachiniSam**
## Purpose
With this pr, a new endpoint is added to get the documentation that is needed for the Statement-editor in low-code-editor.
Related to the issue https://github.com/wso2-enterprise/internal-support-ballerina/issues/73

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/symbol/SymbolDocumentation.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+ */
+public class SymbolDocumentation {
+    private final String description;
+    private final List<ParameterInfo> parameters;
+    private final String returnValueDescription;
+    private final String deprecatedDocumentation;
+    private final List<ParameterInfo> deprecatedParams;
+
+
+    public SymbolDocumentation(Optional<Documentation> documentation, List<ParameterInfo> parameters,
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/symbol/SymbolDocumentation.java:34]: Using optional as a parameter is a bad practice. Instead we can overload the constructor with and without documentation parameter.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/symbol/SymbolDocumentation.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    public List<ParameterInfo> getDeprecatedParams() {
+        return deprecatedParams;
+    }
+
+    /**
+     * Represents a parameter information.
+     */
+    public static class ParameterInfo {
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/symbol/SymbolDocumentation.java:69]: Can we reuse `org.ballerinalang.langserver.signature.Parameter`and `ParameterInfoModel` used in signature help of LS instead?

**Review by IMS94 — CHANGES_REQUESTED** [review]
Can't we use `textDocument/hover` or `HoverUtil` in LS to get the required documentation instead of duplicating a seemingly similar use case?

**Comment by sachiniSam** [comment]
> Can't we use `textDocument/hover` or `HoverUtil` in LS to get the required documentation instead of duplicating a seemingly similar use case? If required, we can have an offline discussion on this.

As discussed offline, one of the main requirement from the low-code-editor is to get the parameter information and to separately render based on the type. (With the Statement editor user will be able to pick on which parameters to be selected).
With  `textDocument/hover` we get this info in a string which is not very scalable to extract each of these details separately. Therefore this endpoint is added to get the details in the required format

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/symbol/BallerinaSymbolService.java [hunk 3]
```diff
@@ -117,6 +128,100 @@ public CompletableFuture<ExpressionTypeResponse> type(ExpressionTypeRequest requ
...
+                                restParameter.paramKind().name(),
+                                CommonUtil.getModifiedTypeName(context, restParameter.typeDescriptor())
+                        ));
+                    }
+                }
+
+                List<SymbolDocumentation.ParameterInfo> deprecatedParams = new ArrayList<>(
+                        documentation.get().deprecatedParametersMap().size());
+                Map<String, String> deprecatedParamMap = documentation.get().deprecatedParametersMap();
+                if (!deprecatedParamMap.isEmpty()) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/symbol/BallerinaSymbolService.java:202]: We can remove this if condition

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/symbol/BallerinaSymbolService.java [hunk 4]
```diff
@@ -117,6 +128,100 @@ public CompletableFuture<ExpressionTypeResponse> type(ExpressionTypeRequest requ
...
+
+                Optional<Documentation> documentation = symbolAtCursor.get() instanceof Documentable ?
+                        ((Documentable) symbolAtCursor.get()).documentation() : Optional.empty();
+
+                if (documentation.isEmpty()) {
+                    return symbolInfoResponse;
+                }
+
+                List<SymbolDocumentation.ParameterInfo> symbolParams = new ArrayList<>(documentation.get().
+                        parameterMap().size());
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/symbol/BallerinaSymbolService.java:171]: What if the documentation of a particular parameter is missing? Wouldn't we get an exception when we try to add new SymbolDocumentation.PramaterInfo entry for that parameter in line 179?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/symbol/BallerinaSymbolService.java [hunk 5]
```diff
@@ -117,6 +128,100 @@ public CompletableFuture<ExpressionTypeResponse> type(ExpressionTypeRequest requ
...
+
+                if (symbolAtCursor.get() instanceof FunctionSymbol) {
+                    FunctionSymbol functionSymbol = (FunctionSymbol) symbolAtCursor.get();
+
+                    if (functionSymbol.typeDescriptor().params().isPresent()) {
+                        List<ParameterSymbol> parameterSymbolList = functionSymbol.typeDescriptor().params().get();
+                        parameterSymbolList.forEach(parameterSymbol -> {
+                            symbolParams.add(new SymbolDocumentation.ParameterInfo(
+                                    parameterSymbol.getName().get(),
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/symbol/BallerinaSymbolService.java:180]: Can there be instances where the name would be empty?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/symbol/BallerinaSymbolService.java [hunk 6]
```diff
@@ -117,6 +128,100 @@ public CompletableFuture<ExpressionTypeResponse> type(ExpressionTypeRequest requ
...
+                                    documentation.get().parameterMap().get(parameterSymbol.getName().get()),
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/symbol/BallerinaSymbolService.java:181]: documentation.get().parameterMap().get(parameterSymbol.getName().get()) can return `null` in some cases. Shall we verify this using a filter

**Review by malinthar — CHANGES_REQUESTED** [review]

**Review by malinthar — DISMISSED** [review]
LGTM!

**Pull request closed without merging**